File: frontend/src/api/faculty.ts
import { apiRequest } from "./client";
import type { Faculty, FacultyInput, FacultyListResponse } from "../types/faculty";

export type FacultySearch = {
  page: number;
  pageSize: number;
  search?: string;
  searchBy: "name" | "employee_number";
};

export function listFaculty(token: string, search: FacultySearch): Promise<FacultyListResponse> {
  const query = new URLSearchParams({ page: String(search.page), page_size: String(search.pageSize) });
  if (search.search?.trim()) query.set(search.searchBy, search.search.trim());
  return apiRequest<FacultyListResponse>(`/faculty?${query.toString()}`, { token });
}

export function getFaculty(token: string, facultyId: string): Promise<Faculty> {
  return apiRequest<Faculty>(`/faculty/${facultyId}`, { token });
}

export function createFaculty(token: string, input: FacultyInput): Promise<Faculty> {
  return apiRequest<Faculty>("/faculty", { method: "POST", token, body: input });
}

export function updateFaculty(token: string, facultyId: string, input: FacultyInput): Promise<Faculty> {
  return apiRequest<Faculty>(`/faculty/${facultyId}`, { method: "PUT", token, body: input });
}

export function deleteFaculty(token: string, facultyId: string): Promise<void> {
  return apiRequest<void>(`/faculty/${facultyId}`, { method: "DELETE", token });
}

File: backend/app/api/v1/faculty.py
"""Faculty API endpoints."""

from uuid import UUID

from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy.orm import Session

from app.db.session import get_db_session
from app.schemas.faculty import FacultyCreate, FacultyListResponse, FacultyResponse, FacultyUpdate
from app.services.faculty import FacultyService

router = APIRouter(prefix="/faculty", tags=["faculty"])
service = FacultyService()


@router.post("", response_model=FacultyResponse, status_code=status.HTTP_201_CREATED)
def create_faculty(
    faculty_data: FacultyCreate, session: Session = Depends(get_db_session)
) -> FacultyResponse:
    return service.create(session, faculty_data)


@router.get("", response_model=FacultyListResponse)
def list_faculty(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    name: str | None = Query(default=None, min_length=1, max_length=200),
    employee_number: str | None = Query(default=None, min_length=1, max_length=30),
    session: Session = Depends(get_db_session),
) -> FacultyListResponse:
    faculty_members, total = service.list(
        session,
        page=page,
        page_size=page_size,
        name=name,
        employee_number=employee_number,
    )
    return FacultyListResponse(items=faculty_members, page=page, page_size=page_size, total=total)


@router.get("/{faculty_id}", response_model=FacultyResponse)
def get_faculty(faculty_id: UUID, session: Session = Depends(get_db_session)) -> FacultyResponse:
    return service.get_or_404(session, faculty_id)


@router.put("/{faculty_id}", response_model=FacultyResponse)
def replace_faculty(
    faculty_id: UUID, faculty_data: FacultyUpdate, session: Session = Depends(get_db_session)
) -> FacultyResponse:
    return service.replace(session, faculty_id, faculty_data)


@router.delete("/{faculty_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_faculty(faculty_id: UUID, session: Session = Depends(get_db_session)) -> Response:
    service.delete(session, faculty_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
